euphoria-hadoop/src/main/java/cz/seznam/euphoria/hadoop/output/HadoopSink.java

```diff
@@ -30,47 +33,40 @@
 import java.io.IOException;
 import java.util.Objects;
 
-public class HadoopSink<K, V>
-    implements DataSink<Pair<K, V>> {
+public class HadoopSink<K, V> implements DataSink<Pair<K, V>> {
 
-  private final Class<? extends OutputFormat<K, V>> hadoopFormatCls;
+  private final Class<? extends OutputFormat<K, V>> outputFormatClass;
   private final SerializableWritable<Configuration> conf;
+  private final SerializableWritable<JobID> jobID;
 
-  private transient OutputFormat<K, V> hadoopFormatInstance;
-
-  public HadoopSink(Class<? extends OutputFormat<K, V>> hadoopFormatCls,
-                    Configuration hadoopConf)
-  {
-    this.hadoopFormatCls = Objects.requireNonNull(hadoopFormatCls);
-    this.conf = new SerializableWritable<>(Objects.requireNonNull(hadoopConf));
+  public HadoopSink(Class<? extends OutputFormat<K, V>> outputFormatClass,
+                    Configuration conf) {
+    this.outputFormatClass = Objects.requireNonNull(outputFormatClass);
+    this.conf = new SerializableWritable<>(Objects.requireNonNull(conf));
+    this.jobID = new SerializableWritable<>(HadoopUtils.getJobID());
   }
 
   @Override
   public void initialize() {
-    try {
-      OutputCommitter committer = getHadoopFormatInstance()
-          .getOutputCommitter(
-              HadoopUtils.createTaskContext(conf.getWritable(), 0));
-
-      committer.setupJob(HadoopUtils.createJobContext(conf.getWritable()));
-    } catch (Exception ex) {
-      throw new RuntimeException(ex);
-    }
+    ExceptionUtils.unchecked(() -> {
+      final TaskAttemptContext setupContext =
+          HadoopUtils.createSetupTaskContext(conf.get(), jobID.get());
+      newOutputFormatInstance()
```

Comment:
I think this violates hadoop OutputFormat's contract. We must call:
 1) checkOutputSpecs()
 2) getRecordWriter()
 3) getOutputCommitter()

All of this in the *same* instance of the OutputFormat. There can be some logic that relies on this and the contract has to be preserved.

Reply:
This means we have to lazy initialize instance of `OutputFormat` in `transient` field and when the instance is created it has to be configured and setup.

Reply:
no, this is where the original problem initiated from, HadoopSink is created once per jvm (for the given sink) and needs to handle more than one task (therefore it cannot be reused).

Reply:
agreed that checkOutputSpecs should be called during initalization phase (in driver)

Reply:
main issues is getOutputCommiter method in FileOutputFormat
```
  public synchronized 
     OutputCommitter getOutputCommitter(TaskAttemptContext context
                                        ) throws IOException {
    if (committer == null) {
      Path output = getOutputPath(context);
      committer = new FileOutputCommitter(output, context);
    }
    return committer;
  }
}
```